extensions/amp-story/0.1/pagination-buttons.js

```diff
@@ -0,0 +1,187 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {dict} from './../../../src/utils/object';
+import {dev} from '../../../src/log';
+import {renderAsElement} from './simple-template';
+import {EventType, dispatch} from './events';
+import {StateChangeType} from './navigation-state';
+
+
+/** @typedef {{className: string, triggers: (string|undefined)}} */
+let ButtonStateDef;
+
+
+/** @const {!Object<string, !ButtonStateDef>} */
+const BackButtonStates = {
+  HIDDEN: {className: 'i-amphtml-story-button-hidden'},
+  PREVIOUS_PAGE: {
+    className: 'i-amphtml-story-back-prev',
+    triggers: EventType.PREVIOUS_PAGE,
+  },
+  CLOSE_BOOKEND: {
+    className: 'i-amphtml-story-back-close-bookend',
+    triggers: EventType.CLOSE_BOOKEND,
+  },
+};
+
+
+/** @const {!Object<string, !ButtonStateDef>} */
+const ForwardButtonStates = {
+  NEXT_PAGE: {
+    className: 'i-amphtml-story-fwd-next',
+    triggers: EventType.NEXT_PAGE,
+  },
+  SHOW_BOOKEND: {
+    className: 'i-amphtml-story-fwd-more',
+    triggers: EventType.SHOW_BOOKEND,
+  },
+  REPLAY: {
+    className: 'i-amphtml-story-fwd-replay',
+    triggers: EventType.REPLAY,
+  },
+};
+
+
+/** @private @const {!./simple-template.ElementDef} */
+const BUTTON = {
+  tag: 'div',
+  attrs: dict({'class': 'i-amphtml-story-button-container'}),
+  children: [
+    {
+      tag: 'button',
+      attrs: dict({'class': 'i-amphtml-story-button-move'}),
+    },
+    {
+      tag: 'div',
+      attrs: dict({'class': 'i-amphtml-story-prev-sentinel'}),
+    },
+  ],
+};
+
+
+/**
+ * @param {!Element} hoverEl
+ * @param {!Element} targetEl
+ * @param {string} className
+ */
+function setClassOnHover(hoverEl, targetEl, className) {
+  hoverEl.addEventListener('mouseenter', () => {
+    targetEl.classList.add(className);
+  });
+  hoverEl.addEventListener('mouseleave', () => {
+    targetEl.classList.remove(className);
+  });
+}
+
+
+/**
+ * @param {!Document} doc
+ * @param {!ButtonState} defaultState
+ * @param {string} className
+ * @return {{element: !Element, setState: function(!ButtonStateDef):void}}
+ */
+function createButton(doc, defaultState, className) {
+  let state = defaultState;
+
+  const element = renderAsElement(doc, BUTTON);
+
+  element.classList.add(className, defaultState.className);
+
+  element.addEventListener('click', e => {
+    if (!state.triggers) {
+      return;
+    }
+    e.preventDefault();
+    dispatch(element, dev().assert(state.triggers), /* opt_bubbles */ true);
+  });
+
+  return {element, setState: newState => {
```

Comment:
Minor: I think this would be easier to read as a new class instead of storing state as a bound var in the closure scope.

```js
class PaginationButton {
  constructor(doc, initialState) {
    this.state_ = initialState;
    this.button = renderAsElement(doc, BUTTON);
  }
  updateState(state) {
    // ...
  }
}
```

Then you could easily inspect the current state of a button in debugger.

Reply:
Done.